### ISSUE TITLE
History.com - Embedded videos [Anti-Adblock]

### ISSUE THREAD

**Issue by THEtomaso** [uBlockOrigin/uAssets#5704]
### Affected sites:
Any site that features embedded videos from History.com (via `akamaihd.net`).

### Issue:
The rule `@@||history.com^$generichide` (found in '**EasyList**') only works for videos shown natively, at `history.com`.
A script injection rule is needed, to cover the videos which are embedded at other sites too.

### Example:
**Embedded video, at TV2.no:**
`https://www.tv2.no/nyheter/10624595/`
(Scroll down to the embedded video, with the headline "**Se video: – Hva i h...... er det der?!**")

**History.com's embedded link (region locked to Norway):**
`https://www.history.com/embed/1521526339504/`
(For testing purposes, the **EasyList** rule needs to be disabled first: `@@||history.com^$generichide,badfilter`)

**History.com's native article, featuring the same video:**
`https://www.history.com/videos/uss-roosevelt-gofast-ufo-declassified-video/`
(For testing purposes, the **EasyList** rule needs to be disabled first: `@@||history.com^$generichide,badfilter`)

### Note:
Since History.com seems to enforce highly specific region restrictions, troubleshooting this problem has proven to be a bit difficult.
Even when accessing the native page above, users might be redirected to a site closer to their location, as experienced by lassekongo83 here:
https://github.com/DandelionSprout/adfilt/issues/7#issuecomment-495987239

### Screenshots (from the TV2.no page):
**Current situation:**
![filterissue-tv2-historyvideo1](https://user-images.githubusercontent.com/9382674/58366864-9c30f280-7ed8-11e9-81ca-f40c55411e4c.jpg)

**After having added `@@||tv2.no^$generichide`:**
![filterissue-tv2-historyvideo2](https://user-images.githubusercontent.com/9382674/58366867-a652f100-7ed8-11e9-8799-75db1aab84a7.jpg)

### Elaboration:
As you can see from the screenshots above, `$generichide` will work for embdedded videos too.
But of course, this solution isn't very practical, as it would have to be used for every site where the videos are shown!

A semi-fix has been posted by DandelionSprout here:
https://github.com/DandelionSprout/adfilt/issues/7#issuecomment-495911260
..but it has some rather serious limitations.
Most importantly, since it removes the **Play** button, it gives the false impression that the embedded videos are nothing more than plain images!

### Reference (outdated fix):
https://github.com/uBlockOrigin/uAssets/issues/804/

--

* OS/version: Windows 8.1 Pro (x64)
* Browser/version: Pale Moon v28.5.0 (x64)
* Adblock Extension/version: uBlock Origin v1.16.4.10 (XUL)

**Comment by ryanbr** [comment]
a $generichide as a badfilter? so anyone not using an adblocker would also be part of this.

**Comment by mapx-** [comment]
> **For testing purposes**, the EasyList rule needs to be disabled first: @@||history.com^$generichide,badfilter

**Comment by ryanbr** [comment]
Anyways, here is a fix;

`@@||amazon-adsystem.com/aax2/apstag.js$script,domain=history.com`

**Comment by THEtomaso** [comment]
That fix doesn't work at all from my end.
Not even for the videos hosted natively at history.com

**Comment by ryanbr** [comment]
Okay; how about this?

```
@@||cloudfront.net^$domain=history.com
@@||amazon-adsystem.com/aax2/apstag.js$script,domain=history.com
```

**Comment by THEtomaso** [comment]
Nope, that doesn't work either.

I've spent a fair amount of time troubleshooting this myself.
If the solution was as simple as whitelisting or blocking a few scripts, I think I would have figured it out by now.

Like I pointed out in the first post; some type of a script injection rule might be needed in this case.

**Comment by ryanbr** [comment]
I can't reproduce it anymore, essentially those 2 whitelists fix the "disable adblock".  tl;dr works for me.

**Comment by THEtomaso** [comment]
Indeed they seem to have a lot of regional restrictions in place, making this whole thing very difficult to reproduce, depending on your location.
If seomeone want to have a go, testing this from my location, you'll find some free Norwegian proxies here:
`http://spys.one/free-proxy-list/NO/`

**Comment by mapx-** [comment]
testing what ? which is the link / page / video to test ?

**Comment by THEtomaso** [comment]
All of which is listed in the first post, of course.

**Comment by mapx-** [comment]
I tested `https://www.history.com/embed/1521526339504` (us ip / norv ip) all fine here, no anti adb message.

**Comment by THEtomaso** [comment]
Tested under which circumstances?
As I've already pointed out; `@@||history.com^$generichide` effectively kills the adblock detection at history.com, but doesn't help for the same videos when they're embedded elsewhere.

**Comment by mapx-** [comment]
this video (the same from the embedded one above) is working fine
`https://www.tv2.no/nyheter/9562173/`

so, you are searching for a general fix, otherwise the single videos can be easily fixed, right ?
I guess there is no general fix and it depends on how the embedded videos are managed by the various hosting pages

**Comment by THEtomaso** [comment]
> so, you are searching for a general fix, otherwise the single videos can be easily fixed, right ?

Yes, `$generichide` could be used for every single site, which features embedded videos from History.com.
My assumption though, is that a single script injection rule should be able to fix the issue for all sites.
..although, I don't know if such a fix is even possible.

**Comment by kulfoon** [comment]
> mapx : this video (the same from the embedded one above) is working fine
> `https://www.tv2.no/nyheter/9562173/`

Because you checked a wrong video.

**Comment by THEtomaso** [comment]
Yeah, he probably did.
It's not the big one (at the top of the page), but the smaller one (further down).
..like stated in my first post:
> Scroll down to the embedded video, with the headline "**Se video: – Hva i h...... er det der?!**"

Nevertheless, I think it has been made quite clear now, what the problem is.

**Comment by gwarser** [comment]
This is because `$generichide` is not applied to iframe. It works in ABP/AdBlock. I think this is not first issue about this. I remember something about streaming site(?), anyone recall? @gorhill is't this a bug? Maybe it should be changed?

**Comment by gorhill** [comment]
> @gorhill is't this a bug?

Seems like it, `generichide` should definitely be applied:

![a](https://user-images.githubusercontent.com/585534/58768721-79a07880-856d-11e9-8d2a-62a3c8c2a70d.png)

**Comment by gorhill** [comment]
To be clear, what I reported above is with 1.19.7rc1 -- I had even failed to notice the bug report was for 1.16.4.10.

So latest uBO is affecting as and I find this baffling that this bug was not spotted before. The `generichide` filter had been evaluated for the root context all this time, while it must be evaluated for the frame context.

**Comment by okiehsch** [comment]
I think it was not spotted before because the filter
`@@||example.com^$generichide`
get's applied to all 3rd-party iframes of `example.com`.
As a result the iframes of 
`example1.com`,  `example2.com` etc are all automatically fixed.

This case is slightly different and unusal, there is an existing filter
`@@||example1.com^$generichide`
but no filter
`@@||example.com^$generichide`.

Furthermore I could not reproduce the issue of the OP because I do not hae a norwegian IP.
I tried with a EU and US Ip.

**Comment by gorhill** [comment]
So I guess I won't consider it an emergency fix but it feels really bad to have this in there, and since at least 1.16.0 apparently. I have a fix for current uBO, not sure if the fix is trivial for legacy.

**Comment by okiehsch** [comment]
The issue that a 3rd-party iframe triggers an anti-adblock warning is quite common on warez/streaming sites but we have always fixed it with a `generichide` filter of the 1st-party  domain because most of those sites use different streaming serives and the `generichide` filter fixes the issue for all of them.

TL;DR I don't think an emergency fix is necessary.

**Comment by gwarser** [comment]
Probably this is this streaming site issue: https://github.com/uBlockOrigin/uAssets/issues/2929#issuecomment-406688308

Don't know how much it differs now, but in uBlock 0.9.5.16 also does not work.

**Comment by gwarser** [comment]
And this is also similar, but with `popup` option: https://github.com/uBlockOrigin/uAssets/issues/5116#issuecomment-471895337

**Comment by gorhill** [comment]
`popup` is a completely different code path and semantic, so if there is an issue with it it needs to be filed as a separate issue.

**Comment by THEtomaso** [comment]
> https://github.com/gorhill/uBlock/commit/2eb9b726a59e4959f40e487d00c595660d41b3fe
> https://github.com/gorhill/uBlock/commit/53f3e413f355e84893dc75230cd48d0385ff70e7

Yes, this fixed the problem!
Great job!
And a special thanks, for taking your time to update the Legacy branch too!! :+1: 

I'm closing this report then.

**Comment by gwarser** [comment]
Top document `generichide` exception does not affect iframes now?

**Comment by gorhill** [comment]
I was never supposed to affect iframes, it's a filter which is to apply in the local context, as per ABP filter syntax.

**Comment by llacb47** [comment]
Anti-adblock is back, see `https://old.reddit.com/r/uBlockOrigin/comments/davuwj/cant_get_around_the_adblock_blocker_on_historycom/`

Test URL: `https://www.history.com/shows/forged-in-fire/season-6/episode-30`

@okiehsch Please add
```
||history.com/vendor/sourcepoint-bootstrap.js
||sourcepoint-mms.history.com^
```

**Comment by mapx-** [comment]
Did you test these filters ?
https://github.com/uBlockOrigin/uAssets/commit/5986a37587f471c89df6549ad20d6d8b24616f13

**Comment by llacb47** [comment]
That is a different type of video @mapx- and the whitelist added in https://github.com/uBlockOrigin/uAssets/commit/5986a37587f471c89df6549ad20d6d8b24616f13 lets through alot of ads and tracking.

I don't get any anti-adblock on that page with or without those filters.

@jspenguin2017 that video will not load in Nano FYI...

**Comment by mapx-** [comment]
How could break the playback 2 exceptions ?

**Comment by llacb47** [comment]
@mapx- ok. I found better filters

Replace `@@*$xhr,domain=history.com` with `@@||mavencoalition.io^$xmlhttprequest`

and add `history.com##.m-header-ad:style(height:2px!important)` for placeholder

**Comment by mapx-** [comment]
the second does not work, I'll keep `history.com#@#.m-in-content-ad-row`

**Comment by mapx-** [comment]
instead for your video are working these same filters ?

**Comment by jspenguin2017** [comment]
@llacb47 Can't reproduce with latest filters. 
![image](https://user-images.githubusercontent.com/7283682/65837726-1541bb80-e2b8-11e9-8f44-9b3bd3270746.png)

**Comment by THEtomaso** [comment]
`https://www.history.com/topics/us-government/before-the-supreme-court-sonia-sotomayor-saved-baseball-video?cmpid=HV_O_Site_H_HP_SOTOMAYOR` will not play in Nano.

Not in uBO either, at my end.
I only get a black screen, both in Cromium-ungoogled and in Pale Moon.

**Comment by mapx-** [comment]
black screen using european ip, I needed us ip to test the issue

**Comment by llacb47** [comment]
I have US IP but I can reproduce https://github.com/uBlockOrigin/uAssets/issues/5704#issuecomment-536333577 only in Nano, @thetomaso are your filter lists updated?

**Comment by llacb47** [comment]
add `history.com##phoenix-balloon-header`

**Comment by THEtomaso** [comment]
> are your filter lists updated?

Of course.

**Comment by llacb47** [comment]
I am puzzled, even in Chrome Guest mode the video doesn't play. This must be a Chromium bug...

For example, on `https://www.history.com/topics/immigration/ellis-island` the main video at the top does not load in/show up. However, the video under the heading `Ellis Island Timeline` loads and plays. 😕

**Comment by jspenguin2017** [comment]
It plays fine for me. I can't support random browsers, I will only test with the official Google Chrome or the official Chromium build for Ubuntu (distributed by Canonical).

**Comment by llacb47** [comment]
I'm testing on the latest Google Chrome.

**Comment by jspenguin2017** [comment]
Could be GeoLock then.

**Comment by llacb47** [comment]
Nope, I'm in US and signed in with cable provider.

**Comment by jspenguin2017** [comment]
This works 100% of the time for me. Try a fresh profile, could be your settings.

**Comment by llacb47** [comment]
Fresh profile still does not work... What is going on
